### PR TITLE
mon/MDSMonitor: plug paxos when maybe manipulating osdmap

### PR DESCRIPTION
**Pull request by batrick**
Instead of tracking where exactly we're plugging PAXOS, just do it unconditionally whenever we modify pending and may change the osdmap. There is no downside to doing so and simplifies the code.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test windows

**Comment by batrick** [comment]
jenkins test windows

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test windows

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test windows

File: src/mon/MDSMonitor.cc [hunk 1]
```diff
@@ -2312,6 +2309,9 @@ void MDSMonitor::tick()
 
   auto &pending = get_pending_fsmap_writeable();
 
+  /* batch any changes to pending with any changes to osdmap */
+  paxos.plug();
```
Comment (vshankar) [inline src/mon/MDSMonitor.cc:2313]: I'm new to this pugging thing - Paxos.h says:

```
  /**
   * true if we want trigger_propose to *not* propose (yet)
   */
  bool plugged = false;
```

That comment is itself confusing for me :)

Could you explain a bit about paxos (un)plug or if its explained anywhere @batrick?

Comment (batrick) [reply src/mon/MDSMonitor.cc:2313]: It's a relatively new thing. Its purpose revolves around preventing various code paths from triggering a paxos commit before all desired pending state is ready. For us, that mostly means preventing early commit for (1) pending_metadata which is fixed in another PR, and (2) any changes to the osdmap that should be committed with the fsmap changes.

Comment (vshankar) [reply src/mon/MDSMonitor.cc:2313]: Got it. Thanks for explaining.

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
fs suite run: https://pulpito.ceph.com/vshankar-2023-05-08_02:52:50-fs-wip-vshankar-testing-20230507.032515-testing-default-smithi/

**Comment by vshankar** [comment]
@batrick This change seems to be causing mon crashes with backtrace like

```
2023-05-08T15:51:33.033 DEBUG:tasks.cephfs.filesystem:Creating filesystem 'backup_fs'
2023-05-08T15:51:33.034 DEBUG:teuthology.orchestra.run.smithi188:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool create backup_fs_metadata --pg_num_min 64
2023-05-08T15:51:33.540 INFO:teuthology.orchestra.run.smithi188.stderr:pool 'backup_fs_metadata' created
2023-05-08T15:51:33.558 DEBUG:teuthology.orchestra.run.smithi188:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool create backup_fs_data 64 --pg_num_min 64 --target_size_ratio 0.9
2023-05-08T15:51:34.557 INFO:teuthology.orchestra.run.smithi188.stderr:pool 'backup_fs_data' created
2023-05-08T15:51:34.577 DEBUG:teuthology.orchestra.run.smithi188:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph fs new backup_fs backup_fs_metadata backup_fs_data
2023-05-08T15:51:35.514 INFO:tasks.ceph.mon.a.smithi188.stderr:terminate called after throwing an instance of 'std::out_of_range'
2023-05-08T15:51:35.514 INFO:tasks.ceph.mon.a.smithi188.stderr:  what():  map::at
2023-05-08T15:51:35.515 INFO:tasks.ceph.mon.a.smithi188.stderr:*** Caught signal (Aborted) **
2023-05-08T15:51:35.515 INFO:tasks.ceph.mon.a.smithi188.stderr: in thread 7f14327f9700 thread_name:safe_timer
2023-05-08T15:51:35.515 INFO:tasks.ceph.mon.a.smithi188.stderr: ceph version 18.0.0-3798-g0ffab5d4 (0ffab5d45a1cb2c5156819bac7489c51d6adf628) reef (dev)
2023-05-08T15:51:35.515 INFO:tasks.ceph.mon.a.smithi188.stderr: 1: /lib64/libpthread.so.0(+0x12ce0) [0x7f1440412ce0]
2023-05-08T15:51:35.516 INFO:tasks.ceph.mon.a.smithi188.stderr: 2: gsignal()
2023-05-08T15:51:35.516 INFO:tasks.ceph.mon.a.smithi188.stderr: 3: abort()
2023-05-08T15:51:35.516 INFO:tasks.ceph.mon.a.smithi188.stderr: 4: /lib64/libstdc++.so.6(+0x9009b) [0x7f143f49009b]
2023-05-08T15:51:35.516 INFO:tasks.ceph.mon.a.smithi188.stderr: 5: /lib64/libstdc++.so.6(+0x9653c) [0x7f143f49653c]
2023-05-08T15:51:35.516 INFO:tasks.ceph.mon.a.smithi188.stderr: 6: /lib64/libstdc++.so.6(+0x96597) [0x7f143f496597]
2023-05-08T15:51:35.516 INFO:tasks.ceph.mon.a.smithi188.stderr: 7: /lib64/libstdc++.so.6(+0x967f8) [0x7f143f4967f8]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 8: /lib64/libstdc++.so.6(+0x9204b) [0x7f143f49204b]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 9: (MDSMonitor::maybe_resize_cluster(FSMap&, int)+0x3cc) [0x56505415ba8c]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 10: (MDSMonitor::tick()+0x18a) [0x56505416a3fa]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 11: (Monitor::tick()+0xc0) [0x565053f16b30]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 12: (Context::complete(int)+0xd) [0x565053f3a0ad]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 13: (CommonSafeTimer<std::mutex>::timer_thread()+0x127) [0x7f144459b307]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 14: (CommonSafeTimerThread<std::mutex>::entry()+0x11) [0x7f144459c431]
2023-05-08T15:51:35.517 INFO:tasks.ceph.mon.a.smithi188.stderr: 15: /lib64/libpthread.so.0(+0x81cf) [0x7f14404081cf]
2023-05-08T15:51:35.518 INFO:tasks.ceph.mon.a.smithi188.stderr: 16: clone()
2023-05-08T15:51:35.518 INFO:tasks.ceph.mon.a.smithi188.stderr:2023-05-08T15:51:35.512+0000 7f14327f9700 -1 *** Caught signal (Aborted) **
2023-05-08T15:51:35.518 INFO:tasks.ceph.mon.a.smithi188.stderr: in thread 7f14327f9700 thread_name:safe_timer
```

See: https://pulpito.ceph.com/vshankar-2023-05-08_02:52:50-fs-wip-vshankar-testing-20230507.032515-testing-default-smithi/7266295/

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/59691

**Comment by batrick** [comment]
Failure is not caused by this PR.

**Comment by vshankar** [comment]
> Failure is not caused by this PR.

OK. This was the _only_ PR related to MDSMonitor in the test branch, so it did seem likely that it could be causing the monitor to crash.

**Comment by batrick** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#22-June-2023

**Pull request merged**